File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.util.*;
import java.io.*;

class Main {
    static ArrayList<ArrayList<Integer>> tree;
    static int[] parents;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        
        tree = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            tree.add(new ArrayList<>());
        }
        
        for (int i = 1; i <= N-1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            tree.get(start).add(end);
            tree.get(end).add(start);
        }
        br.close();
        
        parents = new int[N+1];
        traverse(1, 0);
            
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        for (int i = 2; i < N+1; i++) {
            bw.write(parents[i] + "\n");
        }
        bw.flush();
        bw.close();
    }
    
    private static void traverse(int cur, int parent) {
        parents[cur] = parent;
        
        for (int next : tree.get(cur)) {
            if (next == parent) {
                continue;
            }
            
            traverse(next, cur);
        }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
sys.setrecursionlimit(10**6) # 재귀 한계를 1,000,000까지 늘림

N = int(input())
edges = [list(map(int, input().split())) for _ in range(N-1)]
tree = { i: [] for i in range(1, N+1) }

for edge in edges:
    s, e = edge # start, end
    # 양방향 저장
    tree[s].append(e)
    tree[e].append(s)

parents = [0] * (N+1)
def traverse(cur, parent):
    parents[cur] = parent
    
    for next in tree[cur]:
        if next == parent:
            continue
        
        traverse(next, cur)

traverse(1, 0)        
print(*parents[2:], sep="\n")